hedera-node/src/main/java/com/hedera/services/store/contracts/HederaAbstractWorldUpdater.java

```diff
@@ -90,91 +109,183 @@ public HederaWorldUpdater updater() {
 	}
 
 	@Override
-	protected HederaWorldState wrappedWorldView() {
-		return world;
+	public Optional<WorldUpdater> parentUpdater() {
+		if (hederaWorldState instanceof WorldUpdater) {
+			return Optional.of((WorldUpdater) hederaWorldState);
+		} else {
+			return Optional.empty();
+		}
 	}
 
-	public static class HederaStackedUpdater
-			extends AbstractWorldUpdater<HederaAbstractWorldUpdater, HederaUpdateTrackingAccount> implements HederaWorldUpdater {
+	@Override
+	public void revert() {
+		for (int i = 0; i < numberOfAddressesClaimed; i++) {
+			hederaWorldState.reclaimContractId();
+		}
+		numberOfAddressesClaimed = 0;
+		createdAccounts.clear();
+		updatedAccounts.clear();
+		deletedAccounts.clear();
+	}
+
+	@Override
+	public Collection<HederaUpdateTrackingAccount> getTouchedAccounts() {
+		return updatedAccounts.values();
+	}
+
+	@Override
+	public Collection<Address> getDeletedAccountAddresses() {
+		return deletedAccounts;
+	}
+
+	public Collection<HederaWorldState.WorldStateAccount> getCreatedAccounts() {
+		return createdAccounts.values();
+	}
+
+	protected abstract HederaUpdateTrackingAccount getForMutation(Address address);
+
+	public static class HederaStackedUpdater extends HederaAbstractWorldUpdater implements HederaWorldUpdater {
+
+		private int numberOfAddressesClaimed = 0;
+
+		private final Map<Address, HederaWorldState.WorldStateAccount> createdAccounts = new HashMap<>();
+		private final Map<Address, HederaUpdateTrackingAccount> updatedAccounts = new HashMap<>();
+		private final List<Address> deletedAccounts = new ArrayList<>();
+
+		private final HederaAbstractWorldUpdater parentUpdater;
 
 		HederaStackedUpdater(final HederaAbstractWorldUpdater world) {
-			super(world);
+			super(world.hederaWorldState);
+			this.parentUpdater = world;
+		}
+
+		//returns account if this account exists in Hedera State
+		@Override
+		public HederaWorldState.WorldStateAccount get(Address address) {
+			if (deletedAccounts.contains(address)) {
+				return null;
+			} else if (updatedAccounts.containsKey(address)) {
+				return updatedAccounts.get(address).getWrappedAccount();
+			} else if (createdAccounts.containsKey(address)) {
+				return createdAccounts.get(address);
+			}
+			return parentUpdater.get(address);
+		}
+
+		//returns a modifiable object
+		@Override
+		public EvmAccount getAccount(final Address address) {
+			// We may have updated it already, so check that first.
+			final HederaUpdateTrackingAccount existing = updatedAccounts.get(address);
+			if (existing != null) {
+				return new WrappedEvmAccount(existing);
+			}
+			if (deletedAccounts.contains(address)) {
+				return null;
+			}
+
+			// Otherwise, get it from our wrapped view and create a new update tracker.
+			final HederaWorldState.WorldStateAccount origin = get(address);
```

Comment:
I think that we need ot return `getForMutation` account not the "read" only WorldStateAccount

Reply:
I think we return a 'tracked' version of the read-only WorldStateAccount which is updateable and tracked in the updatedAccounts